Azure: use OIDC for GitHub Actions

coderbyheart opened the pull request:
nRF Asset Tracker for Azure now uses OpenID Connect to authenticate GitHub Actions against Azure using short-lived credentials.

See 
- https://github.com/NordicSemiconductor/asset-tracker-cloud-azure-js/pull/369
- https://github.com/NordicSemiconductor/asset-tracker-cloud-azure-js/pull/370

coderbyheart committed:
feat(azure): update continuous deployment docs

nRF Asset Tracker for Azure now uses OpenID Connect to
authenticate GitHub Actions against Azure using
short-lived credentials.

See https://github.com/NordicSemiconductor/asset-tracker-cloud-azure-js/pull/369

coderbyheart committed:
fix: update KeyVault permission assigment

coderbyheart committed:
feat: use OIDC for conntinuous integration as well

coderbyheart committed:
fix: trim values

peknis commented on docs/azure/ContinuousDeployment.rst:
```diff
@@ -24,60 +24,69 @@ To enable continuous deployment, complete the following steps:
+   - Set the secrets using the GitHub UI:
 
-#. Export the identifier of the subscription which contains the nRF Asset Tracker resources:
+     Set the following `secrets <https://docs.github.com/en/rest/reference/actions#secrets>`_ through the GitHub UI to an `environment <https://docs.github.com/en/actions/reference/environments#creating-an-environment>`_ called ``production`` in your fork of the nRF Asset Tracker for Azure:
```
```suggestion
     Set the following `secrets <https://docs.github.com/en/rest/reference/actions#secrets>`_ to an `environment <https://docs.github.com/en/actions/reference/environments#creating-an-environment>`_ called ``production`` in your fork of the nRF Asset Tracker for Azure:
```
You already have "using the GitHub UI" above.

peknis commented on docs/azure/ContinuousDeployment.rst:
```diff
@@ -24,60 +24,69 @@ To enable continuous deployment, complete the following steps:
 
-   .. parsed-literal::
-      :class: highlight
+     * ``AZURE_CLIENT_ID`` - Store the application (client) ID of the service principal app registration created in step in the above step.
```
```suggestion
     * ``AZURE_CLIENT_ID`` - Store the application (client) ID of the service principal app registration created in the previous step.
```

peknis commented on docs/azure/ContinuousDeployment.rst:
```diff
@@ -24,60 +24,69 @@ To enable continuous deployment, complete the following steps:
-   .. parsed-literal::
-      :class: highlight
+     * ``AZURE_CLIENT_ID`` - Store the application (client) ID of the service principal app registration created in step in the above step.
+     * ``AZURE_TENANT_ID`` - Store the directory (tenant) ID of the service principal app registration created in step in the above step.
```
```suggestion
     * ``AZURE_TENANT_ID`` - Store the directory (tenant) ID of the service principal app registration created in the previous step.
```

peknis commented on docs/azure/ContinuousDeployment.rst:
```diff
@@ -24,60 +24,69 @@ To enable continuous deployment, complete the following steps:
-      :class: highlight
+     * ``AZURE_CLIENT_ID`` - Store the application (client) ID of the service principal app registration created in step in the above step.
+     * ``AZURE_TENANT_ID`` - Store the directory (tenant) ID of the service principal app registration created in step in the above step.
+     * ``AZURE_SUBSCRIPTION_ID`` - Store the ID of the subscription which contains the nRF Asset Tracker resources.
```
```suggestion
     * ``AZURE_SUBSCRIPTION_ID`` - Store the ID of the subscription containing the nRF Asset Tracker resources.
```

peknis commented on docs/azure/ContinuousDeployment.rst:
```diff
@@ -24,60 +24,69 @@ To enable continuous deployment, complete the following steps:
+     * ``AZURE_SUBSCRIPTION_ID`` - Store the ID of the subscription which contains the nRF Asset Tracker resources.
 
-      export SUBSCRIPTION_ID="*subscription id*"
+     Set the following following values from your :file:`.envrc` file as secrets as well:
```
```suggestion
     Set also the following values from your :file:`.envrc` file as secrets:
```

peknis commented on docs/azure/ContinuousDeployment.rst:
```diff
@@ -24,60 +24,69 @@ To enable continuous deployment, complete the following steps:
+     * ``UNWIRED_LABS_API_KEY``
 
-#. Create the CI credentials:
+   - Set the secrets using the GitHub CLI:
```
```suggestion
   - Alternatively, set the secrets using the `GitHub CLI <https://cli.github.com/>`_:
```

peknis commented on docs/azure/ContinuousDeployment.rst:
```diff
@@ -24,60 +24,69 @@ To enable continuous deployment, complete the following steps:
+   - Set the secrets using the GitHub CLI:
 
-   .. code-block:: bash
+     Alternatively, you can use the `GitHub CLI <https://cli.github.com/>`_  with the environment settings from above:
```
Remove this if you agree with my suggestion above.

peknis commented on docs/azure/ContinuousDeployment.rst:
```diff
@@ -24,60 +24,69 @@ To enable continuous deployment, complete the following steps:
 
-Provide the credentials to GitHub Actions
-*****************************************
+#. Grant the application created in step 1 Contributor permissions for your resource group:
```
```suggestion
#. Grant the application created in Step 1 Contributor permissions for your resource group:
```

peknis commented on docs/azure/ContinuousDeployment.rst:
```diff
@@ -24,60 +24,69 @@ To enable continuous deployment, complete the following steps:
-   * ``APP_NAME``
-   * ``B2C_TENANT``
-   * ``APP_REG_CLIENT_ID``
+#. Grant the application created in step 1 "Key Vault Secrets Officer" to the KeyVault:
```
```suggestion
#. Grant the application created in Step 1 "Key Vault Secrets Officer" rights to the KeyVault:
```

umapraseeda replied:
Could we add a label before step 1 and link to to it for Step 1? Something like 
.. _step1_authentication_openid: before line 32 
and referencing step 1 as :ref:`step 1 <step1_authentication_openid>`

peknis commented on docs/azure/ContinuousDeployment.rst:
```diff
@@ -94,3 +103,8 @@ You can see a workflow run of the Continuous Deployment action:
+More information
+****************
+
+You can read more about how GitHub Actions uses OIDC on `About security hardening with OpenID Connect <https://docs.github.com/en/actions/deployment/security-hardening-your-deployments/about-security-hardening-with-openid-connect>`_ in the GitHub Actions documentation.
```
```suggestion
For more details about how GitHub Actions uses OIDC, read `About security hardening with OpenID Connect <https://docs.github.com/en/actions/deployment/security-hardening-your-deployments/about-security-hardening-with-openid-connect>`_ in the GitHub Actions documentation.
```

coderbyheart committed:
fix: need owner role

peknis commented on docs/azure/ContinuousIntegration.rst:
```diff
@@ -151,31 +151,92 @@ Create a secondary tenant (Azure Active Directory B2C)
 
 #. Link this Azure AD B2C tenant to the subscription for CI by following the `Billing guide <https://docs.microsoft.com/en-us/azure/active-directory-b2c/billing#link-an-azure-ad-b2c-tenant-to-a-subscription>`_.
 
```
As a lot of stuff in this piece of documentation is similar with the Continuous Deployment, make the same changes where applicable.

peknis commented on docs/azure/ContinuousIntegration.rst:
```diff
@@ -205,4 +266,9 @@ To run the end-to-end tests against the solution during development, run the fol
 
-   Azure functions allow only one Client ID and Issuer URL in the Active Directory authentication configuration. So, you cannot interact with this instance from the end-to-end tests and the web application since the user flow name differs (``B2C_1_developer`` for end-to-end tests and ``B2C_1_signup_signin`` for the web application) and it is part of the Issuer URL (for example, ``https://${TENANT ...
\ No newline at end of file
+   Azure functions allow only one Client ID and Issuer URL in the Active Directory authentication configuration. So, you cannot interact with this instance from the end-to-end tests and the web application since the user flow name differs (``B2C_1_developer`` for end-to-end tests and ``B2C_1_signup_signin`` for the web application) and it is part of the Issuer URL (for example, ``https://${TENANT ...
```
```suggestion
   Azure functions allow only one Client ID and Issuer URL in the Active Directory authentication configuration. You cannot interact with this instance from the end-to-end tests and the web application, because the user flow names are different (``B2C_1_developer`` for end-to-end tests and ``B2C_1_signup_signin`` for the web application) and it is part of the Issuer URL (for example, ``https://${TENANT_DOMAIN}.b2clogin.com/${TENANT_DOMAIN}.onmicrosoft.com/v2.0/.well-known/openid-configuration?p=B2C_1_developer``).
```

umapraseeda replied:
```suggestion
   Azure functions allow only one Client ID and Issuer URL in the Active Directory authentication configuration.
   So, you cannot interact with this instance from the end-to-end tests and the web application since the user flow name differs (``B2C_1_developer`` for end-to-end tests and ``B2C_1_signup_signin`` for the web application) and it is part of the Issuer URL (for example, ``https://${TENANT_DOMAIN}.b2clogin.com/${TENANT_DOMAIN}.onmicrosoft.com/v2.0/.well-known/openid-configuration?p=B2C_1_developer``).
```

peknis commented on docs/aws/ContinuousIntegration/WebApp.rst:
```diff
@@ -99,7 +99,7 @@ You need to configure AWS credentials as `GitHub environment secrets <https://do
   - Set the secrets using the GitHub CLI:
```
```suggestion
  - Alternatively, set the secrets using the `GitHub CLI <https://cli.github.com/>`_ (before this, create the ``production`` `environment <https://docs.github.com/en/actions/deployment/targeting-different-environments/using-environments-for-deployment>`_ in your repository):
```

peknis commented on docs/aws/ContinuousIntegration/WebApp.rst:
```diff
@@ -99,7 +99,7 @@ You need to configure AWS credentials as `GitHub environment secrets <https://do
   - Set the secrets using the GitHub CLI:
 
     Alternatively, you can use the `GitHub CLI <https://cli.github.com/>`_  with the
-    environment settings from above:
+    environment settings from above (make sure to create the ``production`` `environment <https://docs.github.com/en/actions/deployment/targeting-different-environments/using-environments-for-deployment>`_ in your repository first):
```
Remove this if you agree with my suggestion above.

peknis commented on docs/azure/ContinuousDeployment.rst:
```diff
@@ -24,60 +24,80 @@ To enable continuous deployment, complete the following steps:
+1. Follow the instructions to `Configure a service principal with a Federated Credential to use OIDC based authentication <https://github.com/Azure/login#configure-a-service-principal-with-a-federated-credential-to-use-oidc-based-authentication>`_.
+   Use ``https://nrfassettracker.invalid/cd`` as the name.
+
+   From the command line this can be achieved using:
```
```suggestion
   On the command line, use the following commands:
```

peknis commented on docs/azure/ContinuousDeployment.rst:
```diff
@@ -24,60 +24,80 @@ To enable continuous deployment, complete the following steps:
+      az rest --method POST --uri "https://graph.microsoft.com/beta/applications/${APPLICATION_OBJECT_ID}/federatedIdentityCredentials" --body '{"name":"GitHub Actions","issuer":"https://token.actions.githubusercontent.com","subject":"repo:NordicSemiconductor/asset-tracker-cloud-azure-js:environment:production","description":"Allow GitHub Actions to modify Azure resources","audiences":["api://Azu ...
 
-#. Export the identifier of the subscription which contains the nRF Asset Tracker resources:
+   Make sure to use the organization and repository name of your fork instead of ``NordicSemiconductor/asset-tracker-cloud-azure-js`` in the command above.
```
```suggestion
   Use the organization and repository name of your fork instead of ``NordicSemiconductor/asset-tracker-cloud-azure-js`` in the command.
```

umapraseeda commented on docs/azure/ContinuousDeployment.rst:
```diff
@@ -24,60 +24,80 @@ To enable continuous deployment, complete the following steps:
+     * ``APP_REG_CLIENT_ID``
 
-   .. code-block:: bash
+     If you have enabled the :ref:`azure-unwired-labs-cell-geolocation`, add your API key as a secret as well:
```
```suggestion
     If you have enabled the :ref:`azure-unwired-labs-cell-geolocation`, add your API key ``UNWIRED_LABS_API_KEY`` as a secret as well.
```

umapraseeda commented on docs/azure/ContinuousDeployment.rst:
```diff
@@ -24,60 +24,80 @@ To enable continuous deployment, complete the following steps:
-            "/subscriptions/${SUBSCRIPTION_ID}/resourceGroups/${RESOURCE_GROUP:-nrfassettracker}" \
-         --sdk-auth \
-         > cd-credentials.json
+     * ``UNWIRED_LABS_API_KEY``
```
```suggestion

```

umapraseeda commented on docs/azure/ContinuousDeployment.rst:
```diff
@@ -24,60 +24,80 @@ To enable continuous deployment, complete the following steps:
+   - Set the secrets using the GitHub CLI:
 
-1. Add the following `secrets <https://docs.github.com/en/rest/reference/actions#secrets>`_ to an `environment <https://docs.github.com/en/actions/reference/environments#creating-an-environment>`_ called ``production`` in your fork of the nRF Asset Tracker for Azure:
+     Alternatively, you can use the `GitHub CLI <https://cli.github.com/>`_  with the environment settings from above (make sure to create the ``production`` `environment <https://docs.github.com/en/actions/deployment/targeting-different-environments/using-environments-for-deployment>`_ in your repository first):
```
If we are adding the above suggestion from Pekka,
```suggestion
     You can use the `GitHub CLI <https://cli.github.com/>`_  with the environment settings from above (make sure to create the ``production`` `environment <https://docs.github.com/en/actions/deployment/targeting-different-environments/using-environments-for-deployment>`_ in your repository first):
```

umapraseeda commented on docs/azure/ContinuousDeployment.rst:
```diff
@@ -24,60 +24,80 @@ To enable continuous deployment, complete the following steps:
+       gh secret set B2C_TENANT --env production --body "${B2C_TENANT}"
+       gh secret set APP_REG_CLIENT_ID --env production --body "${APP_REG_CLIENT_ID}"
+
+#. Grant the application created in step 1 Owner permissions for your resource group:
```
Could we add a label before step 1 and link to to it for Step 1? Something like 
.. _step1_authentication_openid: before line 32 
and referencing step 1 as :ref:`step 1 <step1_authentication_openid>`

umapraseeda commented on docs/azure/ContinuousIntegration.rst:
```diff
@@ -151,31 +151,92 @@ Create a secondary tenant (Azure Active Directory B2C)
+
+To enable continuous deployment, complete the following steps:
+
+1. Fork the `nRF Asset Tracker for Azure repository <https://github.com/NordicSemiconductor/asset-tracker-cloud-azure-js>`_.
+
+#. Fork the `nRF Asset Tracker web application repository <https://github.com/NordicSemiconductor/asset-tracker-cloud-app-js>`_.
+
+#. Update the `deploy.webApp.repository <https://github.com/NordicSemiconductor/asset-tracker-cloud-azure-js/blob/fd3777cde331286faf10e481bdf1a30327882008/package.json#L111>`_ in the :file:`package.json` file of your nRF Asset Tracker for Azure fork. It must point to the repository URL of your fork of the nRF Asset Tracker web application.
```
```suggestion
1. Fork the `nRF Asset Tracker for Azure repository <https://github.com/NordicSemiconductor/asset-tracker-cloud-azure-js>`_.
#. Fork the `nRF Asset Tracker web application repository <https://github.com/NordicSemiconductor/asset-tracker-cloud-app-js>`_.
#. Update the `deploy.webApp.repository <https://github.com/NordicSemiconductor/asset-tracker-cloud-azure-js/blob/fd3777cde331286faf10e481bdf1a30327882008/package.json#L111>`_ in the :file:`package.json` file of your nRF Asset Tracker for Azure fork. It must point to the repository URL of your fork of the nRF Asset Tracker web application.
```

umapraseeda commented on docs/azure/ContinuousIntegration.rst:
```diff
@@ -151,31 +151,92 @@ Create a secondary tenant (Azure Active Directory B2C)
+
+   - Set the secrets using the GitHub UI:
+
+     Set the following `secrets <https://docs.github.com/en/rest/reference/actions#secrets>`_ through the GitHub UI to an `environment <https://docs.github.com/en/actions/reference/environments#creating-an-environment>`_ called ``production`` in your fork of the nRF Asset Tracker for Azure:
```
'environment' is defined twice with different link targets (in line 69 also)

umapraseeda commented on docs/azure/ContinuousIntegration.rst:
```diff
@@ -151,31 +151,92 @@ Create a secondary tenant (Azure Active Directory B2C)
+     * ``APP_REG_CLIENT_ID``
 
-#. Fork the `nRF Asset Tracker for Azure project <https://github.com/NordicSemiconductor/asset-tracker-cloud-azure-js>`_.
+     If you have enabled the :ref:`azure-unwired-labs-cell-geolocation`, add your API key as a secret as well:
```
```suggestion
     If you have enabled the :ref:`azure-unwired-labs-cell-geolocation`, add your API key ``UNWIRED_LABS_API_KEY`` as a secret as well.

umapraseeda commented on docs/azure/ContinuousIntegration.rst:
```diff
@@ -151,31 +151,92 @@ Create a secondary tenant (Azure Active Directory B2C)
+     If you have enabled the :ref:`azure-unwired-labs-cell-geolocation`, add your API key as a secret as well:
 
-#. Add the following secrets to an environment called ``ci``:
+     * ``UNWIRED_LABS_API_KEY``
```
```suggestion
```

umapraseeda commented on docs/azure/ContinuousIntegration.rst:
```diff
@@ -151,31 +151,92 @@ Create a secondary tenant (Azure Active Directory B2C)
+       gh secret set B2C_TENANT --env ci --body "${B2C_TENANT}"
+       gh secret set APP_REG_CLIENT_ID --env ci --body "${APP_REG_CLIENT_ID}"
+
+#. Grant the application created in step 1 Owner permissions for your subscription:
```
Suggesting to link to step 1

wiba-nordic commented on docs/azure/ContinuousDeployment.rst:
```diff
@@ -24,60 +24,80 @@ To enable continuous deployment, complete the following steps:
-      az login
+      az ad app create --display-name 'https://nrfassettracker.invalid/cd'
+      export APPLICATION_OBJECT_ID=`az ad app list | jq -r '.[] | select(.displayName=="https://nrfassettracker.invalid/cd") | .id' | tr -d '\n'`
+      az rest --method POST --uri "https://graph.microsoft.com/beta/applications/${APPLICATION_OBJECT_ID}/federatedIdentityCredentials" --body '{"name":"GitHub Actions","issuer":"https://token.actions.githubusercontent.com","subject":"repo:NordicSemiconductor/asset-tracker-cloud-azure-js:environment:production","description":"Allow GitHub Actions to modify Azure resources","audiences":["api://Azu ...
```
```suggestion
      az rest --method POST --uri "https://graph.microsoft.com/beta/applications/${APPLICATION_OBJECT_ID}/federatedIdentityCredentials" --body '{"name":"GitHub Actions","issuer":"https://token.actions.githubusercontent.com","subject":"repo:*organization*/*repository*:environment:production","description":"Allow GitHub Actions to modify Azure resources","audiences":["api://AzureADTokenExchange"]}' 
```
Would using replaceable terms for the command make more sense than telling the user to replace actual ones?

coderbyheart replied:
I was told by tech writers to not use actually replaceble labels anywhere, but this style for parameters.

coderbyheart committed:
Update docs/azure/ContinuousDeployment.rst

Co-authored-by: Pekka Niskanen <73991005+peknis@users.noreply.github.com>

coderbyheart committed:
Update docs/azure/ContinuousDeployment.rst

Co-authored-by: Pekka Niskanen <73991005+peknis@users.noreply.github.com>

coderbyheart committed:
Update docs/azure/ContinuousDeployment.rst

Co-authored-by: Pekka Niskanen <73991005+peknis@users.noreply.github.com>

coderbyheart committed:
Update docs/azure/ContinuousDeployment.rst

Co-authored-by: Pekka Niskanen <73991005+peknis@users.noreply.github.com>

coderbyheart committed:
Update docs/azure/ContinuousDeployment.rst

Co-authored-by: Pekka Niskanen <73991005+peknis@users.noreply.github.com>

coderbyheart committed:
Update docs/azure/ContinuousIntegration.rst

Co-authored-by: Pekka Niskanen <73991005+peknis@users.noreply.github.com>

coderbyheart committed:
Update docs/aws/ContinuousIntegration/WebApp.rst

Co-authored-by: Pekka Niskanen <73991005+peknis@users.noreply.github.com>

coderbyheart committed:
Update docs/azure/ContinuousDeployment.rst

Co-authored-by: Pekka Niskanen <73991005+peknis@users.noreply.github.com>

coderbyheart committed:
Update docs/azure/ContinuousDeployment.rst

Co-authored-by: Pekka Niskanen <73991005+peknis@users.noreply.github.com>

coderbyheart committed:
Update docs/azure/ContinuousDeployment.rst

Co-authored-by: Pekka Niskanen <73991005+peknis@users.noreply.github.com>

coderbyheart committed:
Update docs/azure/ContinuousDeployment.rst

Co-authored-by: Pekka Niskanen <73991005+peknis@users.noreply.github.com>

coderbyheart committed:
Update docs/azure/ContinuousDeployment.rst

Co-authored-by: Uma Praseeda <55196231+umapraseeda@users.noreply.github.com>

coderbyheart committed:
Update docs/azure/ContinuousDeployment.rst

Co-authored-by: Uma Praseeda <55196231+umapraseeda@users.noreply.github.com>

coderbyheart committed:
Update docs/azure/ContinuousDeployment.rst

Co-authored-by: Uma Praseeda <55196231+umapraseeda@users.noreply.github.com>

coderbyheart committed:
Update docs/azure/ContinuousIntegration.rst

Co-authored-by: Uma Praseeda <55196231+umapraseeda@users.noreply.github.com>

coderbyheart committed:
Update docs/azure/ContinuousIntegration.rst

Co-authored-by: Uma Praseeda <55196231+umapraseeda@users.noreply.github.com>

coderbyheart committed:
Update docs/azure/ContinuousIntegration.rst

Co-authored-by: Uma Praseeda <55196231+umapraseeda@users.noreply.github.com>

coderbyheart committed:
Update docs/azure/ContinuousDeployment.rst

Co-authored-by: Pekka Niskanen <73991005+peknis@users.noreply.github.com>

coderbyheart committed:
fix: link to "step 1"

coderbyheart committed:
fix: remove duplicate link names

coderbyheart committed:
ci: use token permissions

wiba-nordic committed:
fix: unify wording between documents

The same step was worded three different ways between three documents.
This fixes all three to match was was in
``docs/azure/ContinuousDeployment.rst``.

Also fixes duplicate link name in
``docs/aws/ContinuousIntegration/WebApp.rst`` to match fix in
the other two instances.

Signed-off-by: Wille Backman <wille.backman@nordicsemi.no>

wiba-nordic commented:
I'm seeing some unintentional whitespace changes from my commit, but it doesn't look like they are bad changes either.

coderbyheart committed:
fix: sentences should be on separate lines
